packages/bff/src/services/toolsService.ts

```diff
@@ -0,0 +1,393 @@
+      params: {
+        clientId: jwt.payload.sub,
+        keyId: jwt.header.kid,
+      },
+      headers: ctx.headers,
+    })
+    .catch(() => undefined);
+
+  if (!client) {
+    return {
+      data: undefined,
+      errors: [
+        clientAssertionPublicKeyNotFound(jwt.header.kid, jwt.payload.sub),
+      ],
+    };
+  }
+
+  const { encodedPem } = await authorizationClient.client.getClientKeyById({
+    headers: ctx.headers,
+    params: {
+      clientId: client.client.id,
+      keyId: jwt.header.kid,
+    },
+  });
+
+  if (!jwt.payload.purposeId) {
+    return {
+      data: undefined,
+      errors: [purposeIdNotFoundInClientAssertion()],
+    };
+  }
+  const purposeId = unsafeBrandId<PurposeId>(jwt.payload.purposeId);
+
+  if (client.client.kind === authorizationApi.ClientKind.enum.API) {
+    return {
+      errors: undefined,
+      data: {
+        clientKind: authorizationApi.ClientKind.enum.API,
+        kid: jwt.header.kid,
+        algorithm: "RS256",
+        publicKey: encodedPem,
+        clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
+        consumerId: unsafeBrandId<TenantId>(client.client.consumerId),
+        purposeId,
+      },
+    };
+  }
+
+  const purpose = await purposeProcessClient.getPurpose({
+    params: { id: purposeId },
+    headers: ctx.headers,
+  });
+
+  const agreement = await retrieveAgreement(
+    agreementProcessClient,
+    purpose.consumerId,
+    purpose.eserviceId,
+    ctx
+  );
+
+  const descriptor = await retrieveDescriptor(
+    catalogProcessClient,
+    agreement.eserviceId,
+    agreement.descriptorId,
+    ctx
+  );
+
+  return {
+    errors: undefined,
+    data: {
+      clientKind: authorizationApi.ClientKind.enum.CONSUMER,
+      clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
+      kid: jwt.header.kid,
+      algorithm: "RS256",
+      publicKey: encodedPem,
+      purposeId,
+      consumerId: unsafeBrandId<TenantId>(client.client.consumerId),
+      agreementId: unsafeBrandId<AgreementId>(agreement.id),
+      eServiceId: unsafeBrandId<EServiceId>(agreement.eserviceId),
+      agreementState: agreementStateToItemState(agreement.state),
+      purposeState: retrievePurposeItemState(purpose),
+      descriptorState: descriptorStateToItemState(descriptor.state),
+    },
+  };
+}
+
+async function retrieveAgreement(
+  agreementClient: AgreementProcessClient,
+  consumerId: string,
+  eserviceId: string,
+  ctx: WithLogger<BffAppContext>
+): Promise<agreementApi.Agreement> {
+  const agreements = await getAllFromPaginated<agreementApi.Agreement>(
+    async (offset, limit) =>
+      await agreementClient.getAgreements({
+        headers: ctx.headers,
+        queries: {
+          offset,
+          limit,
+          consumersIds: [consumerId],
+          eservicesIds: [eserviceId],
+          states: [
+            agreementApi.AgreementState.Values.ACTIVE,
+            agreementApi.AgreementState.Values.SUSPENDED,
+          ],
+        },
+      })
+  );
```

Comment:
You can reuse the `getAllAgreements` function exported by agreementService without redefining here the getAllFromPaginated wrapper etc.

Reply:
The comment was not so clear :) I was referring to something like this, you can reuse the same `getAllAgreements` function: https://vscode.dev/github/pagopa/interop-be-monorepo/blob/main/packages/bff/src/services/agreementService.ts#L538

Reply:
Yes i would use that but the branch is still a lot behind main and i have not access to that function, i think i will wait until the branch gets aligned to port this refactor